pkg/controller/core/localqueue_controller.go

```diff
@@ -146,20 +156,44 @@ func (r *LocalQueueReconciler) Delete(e event.DeleteEvent) bool {
 }
 
 func (r *LocalQueueReconciler) Update(e event.UpdateEvent) bool {
-	q, match := e.ObjectNew.(*kueue.LocalQueue)
-	if !match {
+	oldLq, oldIsLq := e.ObjectOld.(*kueue.LocalQueue)
+	newLq, newIsLq := e.ObjectNew.(*kueue.LocalQueue)
+	if !oldIsLq || !newIsLq {
 		// No need to interact with the queue manager for other objects.
 		return true
 	}
-	log := r.log.WithValues("localQueue", klog.KObj(q))
+	log := r.log.WithValues("localQueue", klog.KObj(newLq))
 	log.V(2).Info("Queue update event")
-	if err := r.queues.UpdateLocalQueue(q); err != nil {
-		log.Error(err, "Failed to update queue in the queueing system")
+
+	oldStopPolicy := ptr.Deref(oldLq.Spec.StopPolicy, kueue.None)
+	newStopPolicy := ptr.Deref(newLq.Spec.StopPolicy, kueue.None)
+
+	if newStopPolicy == oldStopPolicy {
+		if newStopPolicy == kueue.None {
+			if err := r.queues.UpdateLocalQueue(newLq); err != nil {
+				log.Error(err, "Failed to update queue in the queueing system")
+			}
+			if err := r.cache.UpdateLocalQueue(oldLq, newLq); err != nil {
+				log.Error(err, "Failed to update localQueue in the cache")
+			}
+		}
+		return true
 	}
-	oldQ := e.ObjectOld.(*kueue.LocalQueue)
-	if err := r.cache.UpdateLocalQueue(oldQ, q); err != nil {
-		log.Error(err, "Failed to update localQueue in the cache")
+
+	if newStopPolicy == kueue.None {
+		ctx := logr.NewContext(context.Background(), log)
+		if err := r.queues.AddLocalQueue(ctx, newLq); err != nil {
+			log.Error(err, "Failed to add localQueue to the queueing system")
+		}
+		if err := r.cache.AddLocalQueue(newLq); err != nil {
+			log.Error(err, "Failed to add localQueue to the cache")
+		}
+		return true
 	}
+
+	r.queues.DeleteLocalQueue(oldLq)
+	r.cache.DeleteLocalQueue(oldLq)
```

Comment:
Doesn't this immediately release the quota for all the workloads in this localqueue? Even if `stopPolicy == Hold`?

Reply:
Fixed